app/controllers/uploads_controller.rb

```diff
@@ -53,9 +45,35 @@ def show
 
   private
 
+  def alert_messages(loaded_csv)
+    total_rows_count = loaded_csv.ids.length
+    failed_rows = loaded_csv.failed_instances
+    failed_rows_count = failed_rows.length
+    valid_rows = total_rows_count - failed_rows_count
+    validation_warnings = failed_rows.sort { |a, b| a.errors[:row].first.to_i <=> b.errors[:row].first.to_i }
+                                     .map(&:display_errors_with_row)
+    header_warnings = @upload.all_warnings
+
+    successful_messages = [
+      @upload.csv_type + '.csv',
+      'Total data rows: ' + total_rows_count.to_s,
+      'Valid data rows: ' + valid_rows.to_s,
+      'Invalid data rows: ' + failed_rows_count.to_s
+    ]
+
+    flash[:success] = {
+      'The upload succeeded: ': successful_messages
+    }.compact
+
+    flash[:danger] = {
```

Comment:
Since row validation messages will be much longer typically than header warnings wouldn’t it make sense to list those first?

Reply:
Moved it around. Thanks Matt.